pkg/k8s/k8siface_test.go

```diff
@@ -0,0 +1,276 @@
+	Describe("ListGatewayNodes", testListGatewayNodes)
+	Describe("AddGWLabelOnNode", testAddGWLabelOnNode)
+	Describe("RemoveGWLabelFromWorkerNodes", testRemoveGWLabelFromWorkerNodes)
+})
+
+func testRemoveGWLabelFromWorkerNodes() {
+	t := newInterfaceTestDriver()
+
+	BeforeEach(func() {
+		t.nodes = []*corev1.Node{
+			newNode("node-1", map[string]string{k8s.SubmarinerGatewayLabel: "true"}),
+			newNode("node-2", map[string]string{k8s.SubmarinerGatewayLabel: "false"}),
+			newNode("node-3", map[string]string{k8s.SubmarinerGatewayLabel: "true", "foo": "bar"}),
+			newNode("node-4", nil),
+		}
+	})
+
+	It("should remove the label from all nodes", func() {
+		Expect(t.client.RemoveGWLabelFromWorkerNodes()).To(Succeed())
+		t.assertNoLabel(t.nodes[0].Name, k8s.SubmarinerGatewayLabel)
+		t.assertNoLabel(t.nodes[1].Name, k8s.SubmarinerGatewayLabel)
+		t.assertNoLabel(t.nodes[2].Name, k8s.SubmarinerGatewayLabel)
+		t.assertLabel(t.nodes[2].Name, "foo", "bar")
+	})
+
+	Context("on failure", func() {
+		BeforeEach(func() {
+			fake.NewFailingReactorForResource(&t.kubeClient.Fake, "nodes").SetFailOnUpdate(errors.New("fake error"))
+		})
+
+		It("should return an error", func() {
+			Expect(t.client.RemoveGWLabelFromWorkerNodes()).ToNot(Succeed())
+		})
+	})
+}
+
+func testAddGWLabelOnNode() {
+	t := newInterfaceTestDriver()
+
+	BeforeEach(func() {
+		t.nodes = []*corev1.Node{newNode("node", map[string]string{"foo": "bar"})}
+	})
+
+	When("the gateway label isn't present", func() {
+		It("should add it", func() {
+			Expect(t.client.AddGWLabelOnNode("node")).To(Succeed())
+			t.assertLabel(t.nodes[0].Name, k8s.SubmarinerGatewayLabel, "true")
+			t.assertLabel(t.nodes[0].Name, "foo", "bar")
+		})
+	})
+
+	When("the gateway label is set to false", func() {
+		BeforeEach(func() {
+			t.nodes[0].Labels[k8s.SubmarinerGatewayLabel] = "false"
+		})
+
+		It("should set it to true", func() {
+			Expect(t.client.AddGWLabelOnNode(t.nodes[0].Name)).To(Succeed())
+			t.assertLabel(t.nodes[0].Name, k8s.SubmarinerGatewayLabel, "true")
+		})
+	})
+
+	When("the gateway label is already set to true", func() {
+		BeforeEach(func() {
+			t.nodes[0].Labels[k8s.SubmarinerGatewayLabel] = "true"
+		})
+
+		It("should not try to update it", func() {
+			Expect(t.client.AddGWLabelOnNode(t.nodes[0].Name)).To(Succeed())
+
+			actualActions := t.kubeClient.Fake.Actions()
+			for i := range actualActions {
+				if actualActions[i].GetResource().Resource == "nodes" {
+					Expect(actualActions[i].GetVerb()).ToNot(Equal("update"))
+				}
+			}
+		})
+	})
+
+	When("no labels are present", func() {
+		BeforeEach(func() {
+			t.nodes[0].Labels = nil
+		})
+
+		It("should add the gateway label", func() {
+			Expect(t.client.AddGWLabelOnNode(t.nodes[0].Name)).To(Succeed())
+			t.assertLabel(t.nodes[0].Name, k8s.SubmarinerGatewayLabel, "true")
+		})
+	})
+
+	When("the node doesn't exist", func() {
+		BeforeEach(func() {
+			t.nodes = nil
+		})
+
+		It("should not return an error", func() {
```

Comment:
If the node does not exist, the API should ideally return error since Get method would fail isn't it?

Reply:
The admiral `Update` function does not return an error if not found so the behavior elsewhere is to ignore it. I think that makes sense - if the caller wants to update an object then that should mean they know the object exists so a not found error means it was just deleted. We don't want callers to continuously report that and retry.